### PR TITLE
Adding concept of Temporary HP

### PR DESCRIPTION
**Pull request by magroader**
To add temporary HP, currently you long press on the HP+ button. There is nothing to indicate this - I figured I could add a message when you press + when already at full health, but I'm not 100% sure the best way to set it up to only show such a message once per app launch.

Also note this change does not currently prevent going below 0 HP (which doesn't happen in D&D 5E) nor above maximum health (which also doesn't happen in D&D 5E). I am not sure why the app allows going below 0 or above max, but I didn't want to change it without asking first. I did, however, change the visualization when you go above max to make it clear that you're above max.

#424

### CODE REVIEW

**Comment by Jerakin** [comment]
> Also note this change does not currently prevent going below 0 HP (which doesn't happen in D&D 5E) nor above maximum health (which also doesn't happen in D&D 5E)

Caping it can stop someone from playing the way they want, but no one is hindered by it going above/under max/min. It is (in my mind) a lot more likely that someone will be fully hindered to do something by caping the values than it is to stop them to do something by having no limits. In short it follows my thinking of "leave it as flexible as possible" to not cap it. 

___

Love the visualisation improvements!

I think we should also add the same long press interaction to the hp bar. We can then also add a static note about the temp hp there.

File: screens/party/components/meters.lua [hunk 1]
```diff
@@ -127,7 +205,10 @@ end
 
 function M.on_input(action_id, action)
 	for _, b in pairs(active_buttons) do
-		gooey.button(b.node, action_id, action, b.func, b.refresh)
+		local button = gooey.button(b.node, action_id, action, b.func, b.refresh)
+		if b.long_pressed_time then
+			gooey_buttons.set_long_pressed_time(button, 0.5)
+		end
```
[the remark this answers is not in the document]
Comment (magroader) [reply screens/party/components/meters.lua:211]: I thought it was weird too, but I don't understand where else it could go.
The only place I can find where you can actually set this property is after `gooey.button()` has been called. You need to set the `long_pressed_time` property on the return value of `gooey.button()` (which is itself a return calue of `core.instance()` inside `gooey/internal/button.lua`). The only place we are calling `gooey.button()` is here, and its API requires you pass `action_id` and `action`, which only exist in `on_input`. So unless there is some other location I am not aware of where we can set the property and have it be picked up by gooey properly, this seems to be the "right" place to do it. Even though, yes, it is strange.

Another curiosity: there is an "internal" button.lua file that has `function BUTTON.set_long_pressed_time(button, time)` in it. But this is not exposed anywhere in the public API for gooey - I actually hacked it myself by adding `M.set_long_pressed_time` to `gooey_buttons` which is part of the app, not part of gooey. This seemed to be following in a pattern of general utility stuff that happened to know about gooey's internal API.

So, in short, if you can guide me to where this property is actually supposed to be set, I'd happily do that. But as it is, the public API of gooey is seems very unusual to me.

**Comment by magroader** [comment]
Latest adds ability to long press HP bar (and also moves HP bar press logic from party.lua to meters.lua).

File: screens/party/components/meters.lua (hunk 1 above)
[the remark this answers is not in the document]
Comment (Jerakin) [reply screens/party/components/meters.lua:238]: The `long_pressed_time` work directly on the button that `gooey.button` returns, it's probably a very confusing implementation because the person who actually implemented it is always confused https://github.com/britzl/gooey/pull/46 :joy:

The "normal" way of setting up settings and such is to call the button with the id and it will store that for later, so we can do it in the `ini (or create in this case) and it will still have stored that setting for when calling it in `on_input`

```
M.create(nodes, pokemon_id)
	-- setup all the stuff
	for _, b in pairs(active_buttons) do
		gooey.button(b.node).set_long_pressed_time(0.5)
	end
end
```

[the remark this answers is not in the document]
Comment (magroader) [reply screens/party/components/meters.lua:238]: Oh, okay, I think I see - you can just call `gooey.button` even when not in `on_input`, you just don't pass any actions.

File: utils/gooey_buttons.lua [hunk 2]
```diff
@@ -65,6 +65,10 @@ function M.plus_button(button)
 	end
 end
 
+function M.set_long_pressed_time(button, time)
+	button.long_pressed_time = time
+end
+
```
Comment (Jerakin) [inline utils/gooey_buttons.lua:71]: Don't think we need this. We can simply call `gooey.button(node).set_long_pressed_time(time)` directly

Comment (magroader) [reply utils/gooey_buttons.lua:71]: I clearly do not understand how lua "class" equivalents work.

Comment (Jerakin) [reply utils/gooey_buttons.lua]: It returns a `table` (kind of it's own module) that have a function stored on the key `set_long_pressed_time`, was confusing as hell for me too after only have been using Python before starting with Defold.

Comment (magroader) [reply utils/gooey_buttons.lua]: Yeah, I get the concept, I just don't understand how to spot it or what the standards or keywords are for lua.

Like, in C# you'd have something like:
```
namespace Gooey
{
    public class Button
    {
        private double long_pressed_time = 0.0;

        public void set_long_pressed_time(double t)
        {
            long_pressed_time  = t;
        }

// well ACTUALLY this would be a property with a getter/setter but wanted it to match
    }
}
```

and then your usage code would be like:
```
    var myButton = new Gooey.Button();
    ...
    myButton.set_long_pressed_time(0.5);
```

In lua I don't know how to identify when this sort of thing is happening.

Comment (Jerakin) [reply utils/gooey_buttons.lua]: Reading the source code I guess 🙃 [gooey.button](https://github.com/britzl/gooey/blob/master/gooey/internal/button.lua), but if you don't really follow it (even into core.instance) you will have to assume it does so.

I think the API mention that this is how you set the settings to right? If not I should add it to gooeys API

**Comment by Jerakin** [comment]
Looks good now! If it isn't anything else you can think of I say we merge it

File: utils/gooey_buttons.lua (hunk 2 above)
Comment (magroader) [reply utils/gooey_buttons.lua]: I was considering anything under "internal" to not be part of the public API.

**Comment by magroader** [comment]
Sure, we can merge it. I'm not sure which way you wanted to do that (merge commit, squash, or rebase).
